<a href="https://colab.research.google.com/github/Tirta2241720045/Machine_Learning/blob/main/Minggu6/Minggu6_2241720045_TirtaNBP_Tugas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Nama: Tirta Nurrochman Bintang Prawira
# NIM: 2241720045
# Kelas: TI-3A
# No: 27
#Tugas


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

--------------------------------------------------------------------------------


## Tugas 1
- Terdapat dataset mushroom. Berdasarkan dataset yang tersebut, bandingkan peforma antara algoritma Decision Tree dan RandomForest. Gunakan tunning hyperparameter untuk mendapatkan parameter dan akurasi yang terbaik.

Import Library

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

--------------------------------------------------------------------------------
- numpy dan pandas digunakan untuk memanipulasi dan mengolah data.
- train_test_split digunakan untuk membagi dataset menjadi data latih dan data uji.
- GridSearchCV digunakan untuk melakukan pencarian parameter terbaik untuk model dengan metode cross-validation.
- DecisionTreeClassifier dan RandomForestClassifier adalah model yang akan digunakan.
- accuracy_score dan classification_report digunakan untuk mengevaluasi kinerja model.
--------------------------------------------------------------------------------

Persiapan Data

In [3]:
# Memuat dataset
mushroom_data = pd.read_csv('/content/drive/MyDrive/dataset/mushrooms.csv')

# Melihat 5 baris pertama dataset
print(mushroom_data.head())

# Mengganti label string menjadi numerik (jika diperlukan)
mushroom_data['class'] = mushroom_data['class'].map({'e': 1, 'p': 0})  # e: edible, p: poisonous

# Menggunakan One-Hot Encoding untuk fitur kategorikal
mushroom_data = pd.get_dummies(mushroom_data, drop_first=True)

# Memisahkan fitur dan label
X = mushroom_data.drop('class', axis=1)
y = mushroom_data['class']


  class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1       

--------------------------------------------------------------------------------
- Dataset jamur dimuat ke dalam DataFrame menggunakan pd.read_csv.
- head() digunakan untuk menampilkan lima baris pertama dari dataset agar kita bisa melihat strukturnya.
- Label kolom 'class' yang berisi informasi apakah jamur dapat dimakan ('e') atau beracun ('p') diubah menjadi nilai numerik (1 untuk edible dan 0 untuk poisonous).
- One-Hot Encoding digunakan untuk mengubah fitur kategorikal menjadi format numerik, yang diperlukan oleh algoritma pembelajaran mesin.
- Fitur (X) dipisahkan dari label (y), di mana X adalah semua kolom kecuali 'class', dan y adalah kolom 'class'.
--------------------------------------------------------------------------------

Split Data Training dan Testing

In [5]:
# Membagi data menjadi training dan testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

--------------------------------------------------------------------------------
- Dataset dibagi menjadi data latih (70%) dan data uji (30%) menggunakan train_test_split.
- random_state=42 digunakan untuk memastikan hasil pembagian dataset dapat direproduksi.
--------------------------------------------------------------------------------

Training Decision Tree dengan Tuning Hyperparameter

In [6]:
# Menentukan parameter untuk Grid Search
param_grid_dt = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 15, 20],
    'min_samples_split': [2, 5, 10]
}

# Menggunakan GridSearchCV untuk mencari parameter terbaik
grid_search_dt = GridSearchCV(DecisionTreeClassifier(random_state=42), param_grid_dt, cv=5)
grid_search_dt.fit(X_train, y_train)

# Mendapatkan model terbaik dan akurasi
best_dt_model = grid_search_dt.best_estimator_
y_pred_dt = best_dt_model.predict(X_test)
dt_accuracy = accuracy_score(y_test, y_pred_dt)

print("Best Decision Tree Parameters:", grid_search_dt.best_params_)
print("Decision Tree Accuracy:", dt_accuracy)

Best Decision Tree Parameters: {'criterion': 'gini', 'max_depth': None, 'min_samples_split': 2}
Decision Tree Accuracy: 1.0


--------------------------------------------------------------------------------
- Parameter yang akan diuji untuk Decision Tree didefinisikan dalam param_grid_dt, termasuk kriteria pembagian (criterion), kedalaman maksimum pohon (max_depth), dan jumlah minimum sampel untuk membagi node (min_samples_split).
- GridSearchCV digunakan untuk mencari kombinasi parameter terbaik dari model Decision Tree. cv=5 menunjukkan bahwa cross-validation dilakukan dengan 5 lipatan.
- Model terbaik dari hasil pencarian disimpan dalam best_dt_model, dan akurasinya dihitung dengan membandingkan prediksi pada data uji dengan label sebenarnya menggunakan accuracy_score.
- Hasil parameter terbaik dan akurasi dicetak.
--------------------------------------------------------------------------------

Training Random Forest dengan Tuning Hyperparameter

In [7]:
# Menentukan parameter untuk Grid Search
param_grid_rf = {
    'n_estimators': [50, 100, 150],
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 5, 10, 15, 20],
    'min_samples_split': [2, 5, 10]
}

# Menggunakan GridSearchCV untuk mencari parameter terbaik
grid_search_rf = GridSearchCV(RandomForestClassifier(random_state=42), param_grid_rf, cv=5)
grid_search_rf.fit(X_train, y_train)

# Mendapatkan model terbaik dan akurasi
best_rf_model = grid_search_rf.best_estimator_
y_pred_rf = best_rf_model.predict(X_test)
rf_accuracy = accuracy_score(y_test, y_pred_rf)

print("Best Random Forest Parameters:", grid_search_rf.best_params_)
print("Random Forest Accuracy:", rf_accuracy)

Best Random Forest Parameters: {'criterion': 'gini', 'max_depth': None, 'min_samples_split': 2, 'n_estimators': 50}
Random Forest Accuracy: 1.0


--------------------------------------------------------------------------------
- Parameter untuk model Random Forest didefinisikan dalam param_grid_rf, termasuk jumlah estimator (n_estimators), kriteria pembagian, kedalaman maksimum, dan jumlah minimum sampel untuk membagi node.
- Seperti sebelumnya, GridSearchCV digunakan untuk mencari kombinasi parameter terbaik dengan cross-validation.
- Model terbaik dan akurasinya diperoleh dengan cara yang sama seperti pada model Decision Tree, dan hasilnya dicetak.
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
### Kesimpulan
Setelah melakukan pelatihan dan tuning hyperparameter, hasil yang diperoleh adalah:

- **Decision Tree:**
  - **Best Parameters:** `{'criterion': 'gini', 'max_depth': None, 'min_samples_split': 2}`
  - **Accuracy:** 1.0

- **Random Forest:**
  - **Best Parameters:** `{'criterion': 'gini', 'max_depth': None, 'min_samples_split': 2, 'n_estimators': 50}`
  - **Accuracy:** 1.0

Kedua model mencapai akurasi sempurna (1.0), menunjukkan kemampuan mereka dalam mengklasifikasikan dataset jamur dengan sangat baik. Namun, Random Forest lebih robust terhadap overfitting dibandingkan Decision Tree, menjadikannya pilihan yang lebih baik untuk aplikasi di masa mendatang.

--------------------------------------------------------------------------------


## Tugas 2
- Terdapat dataset mushroom. Berdasarkan dataset tersebut, bandingkan peforma antara algoritma Decision Tree dan AdaBoost. Gunakan tunning hyperparameter untuk mendapatkan parameter dan akurasi yang terbaik.



Import Library

In [22]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import LabelEncoder

--------------------------------------------------------------------------------
1. **`import pandas as pd`**: Untuk manipulasi data.
2. **`import numpy as np`**: Untuk komputasi numerik.
3. **`from sklearn.model_selection import train_test_split, GridSearchCV`**: Untuk membagi data dan pencarian hyperparameter.
4. **`from sklearn.tree import DecisionTreeClassifier`**: Untuk klasifikasi dengan pohon keputusan.
5. **`from sklearn.ensemble import AdaBoostClassifier`**: Untuk algoritma boosting.
6. **`from sklearn.metrics import accuracy_score, classification_report`**: Untuk evaluasi model.
7. **`from sklearn.preprocessing import LabelEncoder`**: Untuk mengonversi label kategori.
--------------------------------------------------------------------------------

Persiapan Data

In [13]:
# Memuat dataset
data = pd.read_csv('/content/drive/MyDrive/dataset/mushrooms.csv')

# Cek informasi dataset
print(data.info())

# Mengubah kolom 'class' menjadi numerik menggunakan LabelEncoder
le = LabelEncoder()
data['class'] = le.fit_transform(data['class'])  # Mengubah kolom target 'class' menjadi numerik

# Mengubah kolom kategorikal lainnya menjadi one-hot encoding
data_encoded = pd.get_dummies(data.drop('class', axis=1), drop_first=True)

# Pisahkan fitur dan target
X = data_encoded
y = data['class']


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

--------------------------------------------------------------------------------
1. **Memuat Dataset**: Membaca dataset dari file CSV.

2. **Cek Informasi Dataset**: Menampilkan informasi tentang DataFrame.

3. **Mengubah Kolom 'Class' Menjadi Numerik**: Mengonversi kolom 'class' menjadi format numerik.

4. **One-Hot Encoding**: Mengonversi kolom kategorikal lainnya menjadi one-hot encoding.

5. **Pisahkan Fitur dan Target**: Memisahkan data menjadi fitur (`X`) dan target (`y`).
--------------------------------------------------------------------------------

Split Data Training dan Testing

In [14]:
# Membagi dataset menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

--------------------------------------------------------------------------------
- **`train_test_split`**: Fungsi ini digunakan untuk membagi dataset.
- **`X`**: Merupakan fitur yang telah diproses.
- **`y`**: Merupakan target yang ingin diprediksi.
- **`test_size=0.3`**: Menentukan bahwa 30% dari data akan digunakan sebagai data uji.
- **`random_state=42`**: Menetapkan seed untuk pembagian data sehingga hasilnya dapat direproduksi.
--------------------------------------------------------------------------------

Training Decision Tree dengan Tuning Hyperparameter

In [17]:
# Tuning hyperparameter untuk Decision Tree
dt_params = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None] + list(range(1, 11)),
    'min_samples_split': [2, 5, 10]
}

dt_classifier = DecisionTreeClassifier()
dt_grid_search = GridSearchCV(dt_classifier, dt_params, cv=5, scoring='accuracy')
dt_grid_search.fit(X_train, y_train)

# Menampilkan hasil terbaik
print("Best Decision Tree Parameters:", dt_grid_search.best_params_)
dt_best = dt_grid_search.best_estimator_

# Evaluasi akurasi
dt_y_pred = dt_best.predict(X_test)
dt_accuracy = accuracy_score(y_test, dt_y_pred)
print("Decision Tree Accuracy:", dt_accuracy)

Best Decision Tree Parameters: {'criterion': 'gini', 'max_depth': None, 'min_samples_split': 2}
Decision Tree Accuracy: 1.0


--------------------------------------------------------------------------------
1. **Tuning Hyperparameter**:
   - **`dt_params`**: Parameter untuk tuning Decision Tree (metode pemisahan, kedalaman, dan minimal sampel untuk pembagian).
  
2. **Inisialisasi Model**:
   - **`dt_classifier`**: Membuat model Decision Tree.

3. **Grid Search**:
   - **`dt_grid_search`**: Mencari parameter terbaik menggunakan `GridSearchCV` dengan akurasi sebagai metrik dan 5-fold cross-validation.

4. **Pelatihan Model**:
   - **`dt_grid_search.fit(X_train, y_train)`**: Melatih model dengan data latih.

5. **Hasil Terbaik**:
   - Menampilkan parameter terbaik yang ditemukan dan menyimpan model tersebut dalam **`dt_best`**.

6. **Evaluasi Akurasi**:
   - **`dt_y_pred`**: Prediksi kelas pada data uji.
   - **`dt_accuracy`**: Menghitung dan menampilkan akurasi model.
--------------------------------------------------------------------------------

Training AdaBoost dengan Tuning Hyperparameter

In [21]:
# Definisikan parameter grid untuk tuning
ab_params = {
    'n_estimators': [50, 100, 150],
    'learning_rate': [0.01, 0.1, 1.0],
}

# Mengubah algoritma menjadi SAMME untuk menghindari FutureWarning dan gunakan estimator
ab_classifier = AdaBoostClassifier(estimator=DecisionTreeClassifier(), algorithm='SAMME')

# Menggunakan GridSearchCV untuk mencari parameter terbaik
ab_grid_search = GridSearchCV(ab_classifier, ab_params, cv=5, scoring='accuracy')
ab_grid_search.fit(X_train, y_train)

# Menampilkan hasil terbaik
print(f"Best AdaBoost Parameters: {ab_grid_search.best_params_}")
print(f"AdaBoost Accuracy: {ab_grid_search.best_score_}")


Best AdaBoost Parameters: {'learning_rate': 0.01, 'n_estimators': 50}
AdaBoost Accuracy: 0.9996481970096746


--------------------------------------------------------------------------------
1. Parameter Grid: Mengatur parameter yang akan diuji pada algoritma AdaBoost, seperti jumlah estimators dan learning rate.
2. Inisialisasi AdaBoost: Membuat model AdaBoost dengan Decision Tree sebagai estimator dasar dan algoritma SAMME untuk menghindari peringatan.
3. Grid Search: Mencari kombinasi parameter terbaik menggunakan GridSearchCV dengan 5-fold cross-validation.
4. Pelatihan: Melatih model dengan data latih.
5. Hasil Terbaik: Menampilkan parameter terbaik yang ditemukan dan akurasi model AdaBoost.
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
### Kesimpulan

1. **Decision Tree**:
   - **Parameter Terbaik**: `criterion` Gini, `max_depth` None, `min_samples_split` 2.
   - **Akurasi**: 1.0 (sempurna).

2. **AdaBoost**:
   - **Parameter Terbaik**: `learning_rate` 0.01, `n_estimators` 50.
   - **Akurasi**: 99.96%.

Kedua model menunjukkan performa yang sangat baik, dengan Decision Tree mencapai akurasi sempurna.

--------------------------------------------------------------------------------

## Tugas 3
- Dengan menggunakan dataset diabetes, buatlah ensemble voting dengan algoritma
1. Logistic Regression
2. SVM kernel polynomial
3. Decission Tree
- Anda boleh melakukan eksplorasi dengan melakukan tunning hyperparameter

# Langkah 1: Import Library dan Memuat Dataset

In [26]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

# Memuat dataset
data = pd.read_csv('/content/drive/MyDrive/dataset/diabetes (1).csv')

# Cek informasi dataset
print(data.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None


--------------------------------------------------------------------------------
1. **Impor Library**: Mengimpor pustaka yang diperlukan untuk pengolahan data dan machine learning.
2. **Muat Dataset**: Membaca dataset diabetes dari file CSV.
3. **Cek Informasi**: Menampilkan informasi tentang struktur dan tipe data dalam dataset.
--------------------------------------------------------------------------------

Langkah 2: Preprocessing Data

In [24]:
# Pisahkan fitur dan target
X = data.drop('Outcome', axis=1)  # Asumsikan kolom target bernama 'Outcome'
y = data['Outcome']

# Membagi dataset menjadi data latih dan data uji
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


--------------------------------------------------------------------------------
1. **Pisahkan Fitur dan Target**:
   - `X` berisi semua fitur (kolom) dari dataset kecuali kolom target (`Outcome`).
   - `y` berisi kolom target (`Outcome`).

2. **Bagi Dataset**:
   - Menggunakan `train_test_split` untuk membagi dataset menjadi data latih (70%) dan data uji (30%).
   - `random_state=42` memastikan pembagian yang konsisten setiap kali kode dijalankan.
--------------------------------------------------------------------------------

Langkah 3: Definisikan dan Tuning Model
1. Logistic Regression

In [27]:
# Membangun pipeline untuk menyekal data dan melakukan Logistic Regression
log_reg = make_pipeline(StandardScaler(), LogisticRegression(max_iter=1000))

# Mendefinisikan hyperparameter untuk GridSearch
log_params = {
    'logisticregression__C': [0.01, 0.1, 1, 10, 100],
    'logisticregression__solver': ['liblinear', 'lbfgs']
}

# Melakukan GridSearchCV
log_grid_search = GridSearchCV(log_reg, log_params, cv=5, scoring='accuracy')
log_grid_search.fit(X_train, y_train)

# Menampilkan parameter terbaik
print("Best Logistic Regression Parameters:", log_grid_search.best_params_)

Best Logistic Regression Parameters: {'logisticregression__C': 0.1, 'logisticregression__solver': 'lbfgs'}


--------------------------------------------------------------------------------
1. **Pemisahan Data**: Memisahkan fitur dan target dari dataset.
2. **Pembagian Data**: Membagi data menjadi set latih (70%) dan set uji (30%).
3. **Pipeline**: Membangun pipeline yang menyekal data menggunakan `StandardScaler` dan menerapkan `LogisticRegression`.
4. **Grid Search**: Menggunakan `GridSearchCV` untuk mencari hyperparameter terbaik untuk model regresi logistik.
5. **Output**: Menampilkan parameter terbaik yang ditemukan oleh Grid Search.
--------------------------------------------------------------------------------


2. SVM dengan Kernel Polynomial

In [28]:
# Tuning hyperparameter untuk SVM
svm_params = {
    'C': [0.01, 0.1, 1, 10, 100],
    'degree': [2, 3, 4],
    'kernel': ['poly']
}

svm = SVC()
svm_grid_search = GridSearchCV(svm, svm_params, cv=5, scoring='accuracy')
svm_grid_search.fit(X_train, y_train)

print("Best SVM Parameters:", svm_grid_search.best_params_)


Best SVM Parameters: {'C': 10, 'degree': 2, 'kernel': 'poly'}


--------------------------------------------------------------------------------
1. **Definisi Hyperparameter SVM**: Menentukan parameter yang dioptimalkan untuk model SVM, termasuk `C`, `degree`, dan `kernel`.
2. **Inisialisasi Model**: Membuat objek SVM dengan `SVC()`.
3. **Grid Search**: Menggunakan `GridSearchCV` untuk menemukan parameter terbaik dengan 5-fold cross-validation dan akurasi sebagai metrik.
4. **Pelatihan Model**: Melatih model dengan `svm_grid_search.fit(X_train, y_train)`.
5. **Output**: Menampilkan parameter terbaik dengan `print("Best SVM Parameters:", svm_grid_search.best_params_)`.
--------------------------------------------------------------------------------

3. Decision Tree

In [29]:
# Tuning hyperparameter untuk Decision Tree
dt_params = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None] + list(range(1, 11)),
    'min_samples_split': [2, 5, 10]
}

dt_classifier = DecisionTreeClassifier()
dt_grid_search = GridSearchCV(dt_classifier, dt_params, cv=5, scoring='accuracy')
dt_grid_search.fit(X_train, y_train)

print("Best Decision Tree Parameters:", dt_grid_search.best_params_)

Best Decision Tree Parameters: {'criterion': 'gini', 'max_depth': 5, 'min_samples_split': 10}


--------------------------------------------------------------------------------
1. **Definisi Hyperparameter Decision Tree**: Menentukan parameter yang dioptimalkan untuk model pohon keputusan, termasuk `criterion`, `max_depth`, dan `min_samples_split`.
2. **Inisialisasi Model**: Membuat objek pohon keputusan dengan `DecisionTreeClassifier()`.
3. **Grid Search**: Menggunakan `GridSearchCV` untuk menemukan parameter terbaik dengan 5-fold cross-validation dan akurasi sebagai metrik.
4. **Pelatihan Model**: Melatih model dengan `dt_grid_search.fit(X_train, y_train)`.
5. **Output**: Menampilkan parameter terbaik dengan `print("Best Decision Tree Parameters:", dt_grid_search.best_params_)`.
--------------------------------------------------------------------------------

4. Membuat Ensemble Voting Classifier

In [30]:
# Menggunakan model terbaik dari grid search
best_log_reg = log_grid_search.best_estimator_
best_svm = svm_grid_search.best_estimator_
best_dt = dt_grid_search.best_estimator_

# Membuat ensemble voting classifier
voting_classifier = VotingClassifier(estimators=[
    ('log_reg', best_log_reg),
    ('svm', best_svm),
    ('decision_tree', best_dt)
], voting='hard')

# Melatih model ensemble
voting_classifier.fit(X_train, y_train)

# Evaluasi akurasi
y_pred = voting_classifier.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Ensemble Voting Accuracy:", accuracy)
print(classification_report(y_test, y_pred))

Ensemble Voting Accuracy: 0.7575757575757576
              precision    recall  f1-score   support

           0       0.80      0.84      0.82       151
           1       0.67      0.60      0.63        80

    accuracy                           0.76       231
   macro avg       0.73      0.72      0.73       231
weighted avg       0.75      0.76      0.75       231



--------------------------------------------------------------------------------
1. **Model Terbaik**: Mengambil model terbaik dari hasil grid search untuk regresi logistik, SVM, dan pohon keputusan.

2. **Ensemble Voting Classifier**: Membuat model ensemble menggunakan `VotingClassifier` dengan voting 'hard', yang menggabungkan prediksi dari ketiga model terbaik.

3. **Pelatihan Model Ensemble**: Melatih `voting_classifier` menggunakan data latih dengan `fit(X_train, y_train)`.

4. **Prediksi dan Evaluasi**: Menghitung akurasi model ensemble pada data uji dengan `accuracy_score` dan menampilkan laporan klasifikasi dengan `classification_report`.
--------------------------------------------------------------------------------

--------------------------------------------------------------------------------
### Kesimpulan
1. **Regresi Logistik**:
   - **Parameter Terbaik**: `C = 0.1`, solver `lbfgs`.
   - **Akurasi**: Tinggi, namun tidak disebutkan nilai spesifik.

2. **SVM (Kernel Polinomial)**:
   - **Parameter Terbaik**: `C = 10`, derajat 2.
   - **Akurasi**: Lebih baik dalam menangani data non-linear dibandingkan regresi logistik.

3. **Pohon Keputusan**:
   - **Parameter Terbaik**: Kriteria `gini`, kedalaman maksimum 5, min_samples_split 10.
   - **Akurasi**: Cenderung overfitting jika tidak disetel dengan baik, tetapi parameter ini membantu menyeimbangkan kompleksitas model.

4. **Ensemble Voting Classifier**:
   - **Akurasi**: 75.76%, yang menunjukkan kombinasi model dapat meningkatkan akurasi secara keseluruhan.
   - **Performa Kelas**: Precision untuk kelas negatif (0) adalah 80%, sementara kelas positif (1) 67%, menunjukkan model lebih baik dalam memprediksi kelas negatif.

--------------------------------------------------------------------------------